### PR TITLE
Niganesh/unit and integration tests 2

### PR DESCRIPTION
**Pull request by nithyaganeshng**
Some more unit tests to cover different sections of SDK.

**Type of change:**
- [ ] Feature work
- [ ] Bug fix
- [ ] Documentation
- [ ] Engineering change
- [x] Test
- [ ] Logging/Telemetry

**Risk**:
- [ ] High – Errors could cause MAJOR regression of many scenarios. (Example: new large features or high level infrastructure changes)
- [ ] Medium – Errors could cause regression of 1 or more scenarios. (Example: somewhat complex bug fixes, small new features)
- [x] Small – No issues are expected. (Example: Very small bug fixes, string changes, or configuration settings changes)

**Work Item links**:
Please include here links for this work item, or deferred work, or related work. E.g. if the refactoring is too big to fit in this PR, or the localized strings need to be updated later, please link the TODO work items here.

**Documentation Links**:
Please include here links to any related background documentation for this PR.

### CODE REVIEW

File: PortableIdentityCard-ClientSDK/src/test/java/com/microsoft/portableIdentity/sdk/CardManagerTest.kt [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        assertThat(actualAudience).isEqualTo(expectedAudience)
+    }
+
+    @Test
+    fun `test to save card`() {
+        runBlocking {
+            val actualResult = cardManager.saveCard(portableIdentityCard)
+            assertThat(actualResult is Result.Success).isTrue()
+        }
+    }
```
[the remark this answers is not in the document]
Comment (AndiMiko) [reply PortableIdentityCard-ClientSDK/src/test/java/com/microsoft/portableIdentity/sdk/CardManagerTest.kt]: Yea I agree. It's questionable whether we want to write this test at all then, because it's unclear to me what the purpose is. We are adding overhead and the only gain is a nice codecoverage %. I'm not sure for these scenarios myself, I usually opted to not write a test then - could be wrong -, but we maybe should also get Daniels opinion in the near future, I believe he is very experienced with this :)

File: PortableIdentityCard-ClientSDK/src/test/java/com/microsoft/portableIdentity/sdk/auth/responses/ResponseTest.kt [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    @Test
+    fun `test create receipt`() {
+        val receipts = response.createReceiptsForPresentedCredentials(entityDid, entityName)
+        val expectedReceiptCount = 1
+        assertThat(receipts.size).isEqualTo(expectedReceiptCount)
+        val receipt = receipts.first()
+        assertThat(receipt.cardId).isEqualTo(cardId)
+        assertThat(receipt.entityName).isEqualTo(entityName)
+        assertThat(receipt.entityIdentifier).isEqualTo(entityDid)
+    }
```
[the remark this answers is not in the document]
Comment (AndiMiko) [reply PortableIdentityCard-ClientSDK/src/test/java/com/microsoft/portableIdentity/sdk/auth/responses/ResponseTest.kt]: I see.
For the time checking you could also just do a range check (it's within 5 seconds of current time for example). I think it's too much effort to introduce DI etc to make this simple test case.

Both the time and duplicate type were more examples. I didn't want to create more work for you. It was more meant as general feedback

File: PortableIdentityCard-ClientSDK/src/test/java/com/microsoft/portableIdentity/sdk/CardManagerTest.kt [hunk 3]
```diff
@@ -55,26 +40,48 @@ class CardManagerTest {
...
+
+        coVerify(exactly = 1) {
+            cardManager.createPresentationResponse(presentationRequest)
+            cardManager.sendPresentationResponse(any(), any(), any())
+            cardRepository.sendPresentationResponse(any(), any(), any())
+            cardManager["createAndSaveReceipt"](allAny<PresentationResponse>())
+            presentationResponse.createReceiptsForPresentedCredentials(testEntityDid, testEntityName)
+        }
+        presentationResponse.getCollectedCards()?.size?.let {
+            coVerify(atMost = it) {
```
Comment (AndiMiko) [inline PortableIdentityCard-ClientSDK/src/test/java/com/microsoft/portableIdentity/sdk/CardManagerTest.kt:81]: nit, `exactly`

File: PortableIdentityCard-ClientSDK/src/test/java/com/microsoft/portableIdentity/sdk/CardManagerTest.kt [hunk 4]
```diff
@@ -55,26 +40,48 @@ class CardManagerTest {
...
+                cardManager["saveReceipt"](allAny<Receipt>())
```
Comment (AndiMiko) [inline PortableIdentityCard-ClientSDK/src/test/java/com/microsoft/portableIdentity/sdk/CardManagerTest.kt:82]: wow I actually wasn't aware mockk can verify private method calls. That's pretty cool. I have had situations where a private part was so complicated that I wanted to mock/verify it but that's often times a code smell. 

While it may be an edge case where you want to do this in general you should not be molding the exact implementation into the test scenario. What this leads to is that every change in code would require a change in test. The test should as abstract as possible to the concrete implementation and as concrete as possible to the inputs/outputs.

This means that rather than listening to inner private methods you should listen to it's outputs. The equivalent here is `picRepository.insert(receipt)`. Sorry I may have misguided you with my earlier comment where I said "saveReceipt". The outcome in this case is pretty much the same. But I hope you get the idea.

File: PortableIdentityCard-ClientSDK/src/test/java/com/microsoft/portableIdentity/sdk/CardManagerTest.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (nithyaganeshng) [reply PortableIdentityCard-ClientSDK/src/test/java/com/microsoft/portableIdentity/sdk/CardManagerTest.kt]: My thought is it doesn't add value now. But if add more logic to that method in the future like do some sort of transformation to the object saved which causes regression we would catch it. This is only for future, currently it has only happy path. I am adding tests for some simple methods like this which has very thin or no logic in the hope of catching future regressions but not 100% sure if that is required right now.

File: PortableIdentityCard-ClientSDK/src/test/java/com/microsoft/portableIdentity/sdk/auth/responses/ResponseTest.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (nithyaganeshng) [reply PortableIdentityCard-ClientSDK/src/test/java/com/microsoft/portableIdentity/sdk/auth/responses/ResponseTest.kt]: I will make the change to test time difference. I was suggesting DI as it is setting time is mixed with creating receipt now which is generally hard to test and also if different implementation need to be used to compute time. But both are not priority now.
